WNS(Windows Notification Service) support

AdamBolfik opened the pull request:
I added the ability to send toast/tile/badge/raw notifications via WNS (Windows Notification Service). 

I have tested toast and tile notifications, but not badge or raw. I know badge should JUST work because it follows the toast and tile format closely. 

As for raw notifications: We are developing our app with ionic/cordova so I haven't even begun to see what raw notifications are even for. The docs say it's just a string passed to the app in the background so I made the methods handle this scenario. But again, I'm just assuming it works/works this way.

There is also some functionality that I haven't added, that would be nice:
* It looks as though WNS can get active devices back with the correct header passed <- use for pruning
* Each call to a create a push notification calls WNS for authentication <- This should be saved until expiration (included in response) to minimize calls.

I've added a little on the README, but it's not even close to complete. Thought that I would send this up for review as it is already massive.

jleclanche commented on push_notifications/admin.py:
```diff
@@ -2,18 +2,17 @@
+	list_filter = ("active",)
 	actions = ("send_message", "send_bulk_message", "prune_devices", "enable", "disable")
-	raw_id_fields = ("user", )
+	raw_id_fields = ("user",)
```
You're making unnecessary whitespace changes all over the place - these pollute the diffs and make it very hard to review the PR. Please don't do that.

AdamBolfik replied:
Sorry,  my ide went crazy on it.

jleclanche commented on push_notifications/wns.py:
```diff
@@ -0,0 +1,359 @@
+		elif err.code == 500:
+			raise WNSNotificationResponseError("HTTP 500: An internal failure caused notification delivery to fail.")
+		elif err.code == 503:
+			raise WNSNotificationResponseError("HTTP 503: The server is currently unavailable.")
```
There is really no need for this. You can `raise WNSNotificationResponseError(err)` instead.

AdamBolfik replied:
Yeah,  figured I would just make a more specific message. I mapped the errors to the wns docs.

jleclanche commented on push_notifications/wns.py:
```diff
@@ -0,0 +1,359 @@
+						},
+					},
+				},
+			}
```
Please don't format that dict like that, it's unreadable. Same comment applies to the tests.

AdamBolfik replied:
K, my ide didn't show sick large indents,  I'll look into it.

AdamBolfik replied:
Such*

AdamBolfik replied:
Alright, I've looked into it... I'm not sure how to format this correctly with tabs. What's your preference? Just put it all one line-ish?

jleclanche commented:
Overall I like the code. The commit log can't stay like that though (see [CONTRIBUTING.md](https://github.com/jleclanche/django-push-notifications/blob/master/CONTRIBUTING.md)). If you don't know how to rebase it, I can do it.

I'll see if someone else can review/test it as well.

AdamBolfik commented:
Sorry, I noticed your contribution log after I started contributing...sigh.

If rebasing is something you do with ease, I'll gladly let you handle it. If not, I can take a look at it tomorrow.

jleclanche commented:
Well, it's a good skill to know regardless, so you can try:

https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History

You don't need to create a new PR. Force-pushing will update this one. If you get stuck/mess up completely, `git reset --hard 6cfcd32` will get you back to where you started :)

AdamBolfik commented:
Alright, I've rebased it. I just squashed the last couple commits about tabs/spaces as my other commits all seem to adhere to 'one change', at least, in my view.

jamaalscarlett commented:
It looks like the whitespace changes are still there.

AdamBolfik commented:
The whitespace keeps you on your toes. ;)

jamaalscarlett commented:
I've got two kids, I'm always on my toes.

jleclanche commented on push_notifications/wns.py:
```diff
@@ -0,0 +1,359 @@
+from . import NotificationError
+from .settings import PUSH_NOTIFICATIONS_SETTINGS as SETTINGS
+
+WNS_ACCESS_URL = "https://login.live.com/accesstoken.srf"
```
This should be a setting (cf. gcm/apns)

AdamBolfik replied:
I'm fine with this either way, but what is the reason for making this configurable at the project level? 
It would seem that if WNS changes the url or api in any way it would have to be maintained by the app(you) and wouldn't be something the project developer would want/need to have knowledge about. 

Basically, If it needs to be overridden, shouldn't it be a bug that needs fixing in master?

jleclanche replied:
It's useful if the upstream URL changes and the DPN project is unmaintained at that time, and/or someone runs a self-hosted WNS dispatch server at a different address.

Both are unlikely, but it mirrors the behaviour in GCM/APNS.

AdamBolfik commented:
Also, this links back to #289 whenever this PR gets closed. Not sure if I needed to link it both ways.

jleclanche commented on push_notifications/wns.py:
```diff
@@ -0,0 +1,359 @@
+Windows Notification Service
+
+Documentation is available on the Windows Dev Center:
+https://msdn.microsoft.com/en-us/windows/uwp/controls-and-patterns/
```
This URL is 404. What's the correct one?

jleclanche commented:
I'm going to touch up the commits and finish off the rebase locally but I would really like at least 1-2 people to tell me they tried it and it works fine.

I notice there's no migration for the WNSDevice model at the very least so I suspect it's not quite complete yet.

AdamBolfik commented on push_notifications/migrations/0003_wnsdevice.py:
```diff
@@ -0,0 +1,33 @@
+# -*- coding: utf-8 -*-
```
Here is the migration

jleclanche commented:
Anyone testing it?

jamaalscarlett commented:
@jleclanche I think it is going to be hard to find people both using this library and using windows phone.

jleclanche commented:
@jamaalscarlett Could you just run it to make sure it doesn't break existing installations? I'm mostly concerned about that.

jamaalscarlett commented:
Will do.  I'm @ work right now.  But I will run it when I get home tonight.

jleclanche commented:
@jamaalscarlett ping

jamaalscarlett commented:
The branch does not through any errors.

jleclanche commented:
Squashed, cleaned up and landed.
